src/display/editor/tools.js

```diff
@@ -977,20 +990,30 @@ class AnnotationEditorUIManager {
         ? anchorNode.parentElement
         : anchorNode;
     if (!anchorElement.closest(".textLayer")) {
-      if (this.#hasSelection) {
-        this.#hasSelection = false;
+      if (this.#selectedTextNode) {
         this.#selectedTextNode = null;
         this.#dispatchUpdateStates({
           hasSelectedText: false,
         });
       }
       return;
     }
-    this.#hasSelection = true;
     this.#selectedTextNode = anchorNode;
     this.#dispatchUpdateStates({
       hasSelectedText: true,
     });
+
+    if (this.#mode !== AnnotationEditorType.HIGHLIGHT) {
+      return;
+    }
+    if (this.isShiftKeyDown) {
+      this.#highlightWhenShiftUp = true;
+    } else {
+      this.#highlightWhenShiftUp = false;
+      window.addEventListener("pointerup", this.highlightSelection.bind(this), {
```

Comment:
While this is effectively just code that you moved/refactored, isn't there a pre-existing issue here?
It seems to me that situation *similar* to the one outlined in https://github.com/mozilla/pdf.js/pull/17650#discussion_r1484821600 would apply here as well?